### ISSUE TITLE
java agent make java thread to be locked

### ISSUE THREAD

**Issue by suppix** [elastic/apm-agent-java#1839]
I have some strange behavior when I use app java agent with my application, 
sometimes threads are in a locked state. 

```
Name: catalina-wsm-exec-62
State: BLOCKED on java.lang.Class@2437c6dc owned by: catalina-wsm-exec-44
Total blocked: 4  Total waited: 11

Stack trace: 
co.elastic.apm.agent.bci.ElasticApmAgent.ensureInstrumented(ElasticApmAgent.java:700)
co.elastic.apm.agent.bci.DynamicTransformerImpl.ensureInstrumented(DynamicTransformerImpl.java:36)
co.elastic.apm.agent.concurrent.JavaConcurrent.captureContext(JavaConcurrent.java:113)
co.elastic.apm.agent.concurrent.JavaConcurrent.withContext(JavaConcurrent.java:136)
co.elastic.apm.agent.concurrent.ExecutorInstrumentation$ExecutorCallableInstrumentation.onSubmit(ExecutorInstrumentation.java:156)
java.lang.invoke.LambdaForm$DMH/2059904228.invokeStatic_LL_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/507911745.linkToTargetMethod(LambdaForm$MH)
java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:132)
com.zaxxer.hikari.pool.HikariPool.addBagItem(HikariPool.java:339)
com.zaxxer.hikari.util.ConcurrentBag.borrow(ConcurrentBag.java:140)
com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:180)
com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:162)
com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:128)
org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource.getConnection(AbstractRoutingDataSource.java:148)
org.springframework.orm.hibernate3.LocalDataSourceConnectionProvider.getConnection(LocalDataSourceConnectionProvider.java:81)
org.hibernate.jdbc.ConnectionManager.openConnection(ConnectionManager.java:446)
org.hibernate.jdbc.ConnectionManager.getConnection(ConnectionManager.java:167)
org.hibernate.jdbc.JDBCContext.connection(JDBCContext.java:160)
org.hibernate.transaction.JDBCTransaction.begin(JDBCTransaction.java:81)
org.hibernate.impl.SessionImpl.beginTransaction(SessionImpl.java:1473)
org.springframework.orm.hibernate3.HibernateTransactionManager.doBegin(HibernateTransactionManager.java:555)
org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:371)
org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:335)
org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:105)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:80)
pro.finance.wsm.aop.ReadOnlyConnectionInterceptor.proceed(ReadOnlyConnectionInterceptor.java:37)
sun.reflect.GeneratedMethodAccessor550.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:621)
org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:610)
org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:65)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:161)
org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:89)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
com.sun.proxy.$Proxy329.findSystemAccountByLogin(Unknown Source)
pro.finance.wsm.facade.impl.MerchantSecurity.getSystemAccountByLogin(MerchantSecurity.java:107)
pro.finance.wsm.aop.MerchantSecurityAspect.isValidCall(MerchantSecurityAspect.java:121)
pro.finance.wsm.aop.MerchantSecurityAspect.securityCheck(MerchantSecurityAspect.java:72)
sun.reflect.GeneratedMethodAccessor542.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:621)
org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:610)
org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:65)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:89)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
com.sun.proxy.$Proxy416.findPaymentByOrderId(Unknown Source)
sun.reflect.GeneratedMethodAccessor540.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:173)
org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:89)
org.apache.cxf.jaxws.JAXWSMethodInvoker.invoke(JAXWSMethodInvoker.java:60)
org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:75)
org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:58)
java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
java.util.concurrent.FutureTask.run(FutureTask.java:256)
org.apache.cxf.workqueue.SynchronousExecutor.execute(SynchronousExecutor.java:37)
org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:106)
   - locked org.apache.cxf.interceptor.ServiceInvokerInterceptor$2@4facfcf
org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:243)
   - locked org.apache.cxf.phase.PhaseInterceptorChain@1d5b4c5e
org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:109)
org.apache.cxf.transport.servlet.ServletDestination.invoke(ServletDestination.java:98)
org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:406)
org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:178)
org.apache.cxf.transport.servlet.AbstractCXFServlet.invoke(AbstractCXFServlet.java:142)
org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:179)
org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:103)
javax.servlet.http.HttpServlet.service(HttpServlet.java:648)
org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:159)
org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:230)
org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
pro.finance.wsm.filters.GetRealIPFilter.doFilter(GetRealIPFilter.java:46)
org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:108)
org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:472)
org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
org.apache.catalina.valves.RequestFilterValve.process(RequestFilterValve.java:319)
org.apache.catalina.valves.RemoteAddrValve.invoke(RemoteAddrValve.java:88)
org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:349)
org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:784)
org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:802)
org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1410)
org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
   - locked org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@c3a569e
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
java.lang.Thread.run(Thread.java:745)
```

**Comment by eyalkoren** [comment]
Thanks for reporting.
What is the effect on your application? Does it get hang (requests are unable to complete)? Is it slowed down?

`BLOCKED` in itself is a valid state, indicating that a thread is waiting for a lock to become available. It could be an indication for a deadlock in some cases, or for high contention, or indicate nothing (normal execution).
Please provide the full thread dump and a description of the effect on your application.
Also, please provide as many details as possible about your application and your agent setup.

**Comment by SylvainJuge** [comment]
Hi @suppix, 

This looks quite similar to https://github.com/elastic/apm-agent-java/issues/1884

Can you tell us which version of the agent was causing this issue ?
This might be something that we already fixed with https://github.com/elastic/apm-agent-java/pull/1758 which is included in version `1.23.0`.

If you are not yet using the latest version of the agent, can you upgrade and tell us if that makes a difference ?
Also, can you tell us what is the relative frequency of that issue ? Is it reproducible with 100% / 50% / less of the times that the application starts (where I guess it happens).

**Comment by SylvainJuge** [comment]
Updating the agent version in #1884 seems to have resolved this issue (it was previously using version `1.18.0`), thus I'll close this issue for now. Please re-open it (or open another one) if it still happens after updating to `1.23.0` or later.
